### PR TITLE
OKTA-418888 Implement Android App Link binding method POC

### PR DESCRIPTION
**Pull request by yannongli-okta**
## Description:

This is the initial PR for implementing Android App Link in SIW so that team can test it on the client side. More features and testing regarding the Factor Challenge flow will be added.

## PR Checklist

- [x] Have you verified the basic functionality for this change?

- [x] Added unit tests?

- [x] Added e2e tests

- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?

### Screenshot/Video:
UX Mocks: https://www.figma.com/file/prXV4T8nj9fy9CwSMGmwcx/OV-OIE_21?node-id=5721%3A55446

Loopback Fallback to App Link:
siw1

App Link Polling Page:
siw2

Complete e2e flow with monolith(please ignore the outdated text as they were recorded at the beginning):
https://okta.box.com/s/d2l31z5jqte9mqckav0kgjydnkabtlni

### Reviewers:

@okta/device-platform @jhe-okta @pradeepdewda-okta @sherizada-okta 

### Issue:

- [OKTA-418888](https://oktainc.atlassian.net/browse/OKTA-418888)

### CODE REVIEW

File: src/v2/view-builder/utils/ChallengeViewUtil.js [hunk 1]
```diff
@@ -83,5 +83,24 @@ export function doChallenge(view) {
         Util.redirect(deviceChallenge.href);
       }
     }));
+    break;
+  case Enums.APP_LINK_CHALLENGE:
+    view.title = loc('appLink.title', 'login');
+    view.add(View.extend({
+      className: 'app-link-content',
+      template: hbs`
+            {{{i18n code="appLink.content" bundle="login"}}}
```
Comment (jhe-okta) [inline src/v2/view-builder/utils/ChallengeViewUtil.js:92]: Use `{{` to escape the content.

**Review by jhe-okta — APPROVED** [review]
Looks good!

File: packages/@okta/i18n/src/properties/login_ok_PL.properties [hunk 2]
```diff
@@ -612,7 +615,7 @@ oie.password.challenge.title = 》Ṽéŕîƒý ŵîţĥ ýöûŕ þåššŵöŕ
 oie.password.enroll.title = 》Šéţ ûþ þåššŵöŕð ฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
 oie.password.passwordLabel = 》Éñţéŕ þåššŵöŕð โ⾼ئ䀕ヸ€홝한Ӝฐโ《
 oie.password.confirmPasswordLabel = 》Ŕé‐éñţéŕ þåššŵöŕð 한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
-oie.password.reset.verification = 》Ṽéŕîƒý ŵîţĥ öñé öƒ ţĥé ƒöļļöŵîñĝ åûţĥéñţîçåţöŕš ţö ŕéšéţ ýöûŕ þåššŵöŕð· 한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
+oie.password.reset.verification = 》Ṽéŕîƒý ŵîţĥ öñé öƒ ţĥé ƒöļļöŵîñĝ šéçûŕîţý ɱéţĥöðš ţö ŕéšéţ ýöûŕ þåššŵöŕð· 홝한Ӝฐโ⾼ئ䀕ヸ€홝한Ӝฐโ《
```
Comment (jmelberg-okta) [inline packages/@okta/i18n/src/properties/login_ok_PL.properties:618]: These were altered as part of #2080, but the dev (me) didn't run the pseudo-loc task. These are **safe** and **valid** changes.

File: src/v2/view-builder/utils/ChallengeViewUtil.js (hunk 1 above)
Comment (jmelberg-okta) [reply src/v2/view-builder/utils/ChallengeViewUtil.js:92]: +1. Using triple curly braces prevents Handlebars from auto-escaping. Please use double bracket notation here:

```suggestion
            {{i18n code="appLink.content" bundle="login"}}
```

File: src/v2/view-builder/utils/ChallengeViewUtil.js [hunk 3]
```diff
@@ -83,5 +83,24 @@ export function doChallenge(view) {
         Util.redirect(deviceChallenge.href);
       }
     }));
+    break;
+  case Enums.APP_LINK_CHALLENGE:
+    view.title = loc('appLink.title', 'login');
+    view.add(View.extend({
+      className: 'app-link-content',
```
Comment (jmelberg-okta) [inline src/v2/view-builder/utils/ChallengeViewUtil.js:90]: Curious why we need a `className` here? Seems like we can simplify this to:

```suggestion
    view.add(loc('appLink.content', 'login'));
```

**Review by jmelberg-okta — APPROVED** [review]

File: src/v2/view-builder/utils/ChallengeViewUtil.js (hunk 3 above)
Comment (yannongli-okta) [reply src/v2/view-builder/utils/ChallengeViewUtil.js:90]: i'm following https://github.com/okta/okta-signin-widget/blob/master/src/v2/view-builder/utils/ChallengeViewUtil.js#L74 as they are very similar. While i didn't find specific use case for the classname, it's being used in tests https://github.com/okta/okta-signin-widget/blob/master/test/unit/spec/v2/view-builder/utils/ChallengeViewUtil_spec.js#L118 probably @jhe-okta has a better idea

**Review by aarongranick-okta — CHANGES_REQUESTED** [review]
please rebase against `master`

**Review by aarongranick-okta — APPROVED** [review]

File: test/unit/spec/v2/view-builder/utils/ChallengeViewUtil_spec.js [hunk 4]
```diff
@@ -136,4 +136,48 @@ describe('v2/utils/ChallengeViewUtil', function() {
...
+    expect(testView.add).toHaveBeenCalledTimes(2);
+    expect(expectedAddArgs[0].className).toBe('app-link-content');
+    expect(expectedAddArgs[0].template.call()).toBe(hbs`
+            {{{i18n code="appLink.content" bundle="login"}}}
+          `.call());
+    let expectedCreateButton = createButton({
+      className: 'al-button button button-wide button-primary',
+      title: loc('oktaVerify.reopen.button', 'login'),
+      click: () => {
+        // only window.location.href can open universal link in iOS/MacOS
```
Comment (sherizada-okta) [inline test/unit/spec/v2/view-builder/utils/ChallengeViewUtil_spec.js:165]: And new devices of Android.

**Pull request closed without merging**